source/interact-data/transaction.txt

```diff
@@ -0,0 +1,283 @@
+   :dedent:
+
+.. note::
+   
+   When you call the ``transaction()`` method on the ``{+odm+}`` module,
+   {+odm+} creates the transaction by using the ``:default`` client.
+
+Ending Transactions
+~~~~~~~~~~~~~~~~~~~
+
+Any exception raised inside the transaction method block ends the
+transaction and rolls back changes. Usually, {+odm+} displays the
+exception, except for the ``Mongoid::Errors::Rollback`` exception. Implement
+this exception in your application to explicitly end the transaction
+without passing on an exception.
+
+Callbacks
+~~~~~~~~~
+
+This transaction API introduces the ``after_commit`` and
+``after_rollback`` callbacks.
+
+{+odm+} triggers the ``after_commit`` callback for an object that was
+created, saved, or deleted in the following cases:
+
+- After the transaction is committed if the object was modified inside
+  the transaction.
+
+- After the object is persisted if the object was modified before
+  the transaction block.
+
+The ``after_commit`` callback is triggered only after all
+other callbacks are executed successfully. Therefore, if an object is
+modified without a transaction, it is possible that the object was
+persisted, but the ``after_commit`` callback was not triggered, for
+example, if {+odm+} raised an exception in the ``after_save`` callback.
+
+The ``after_rollback`` callback is triggered for an object that was
+created, saved, or deleted inside a transaction if the transaction was
+ended. {+odm+} never triggers ``after_rollback`` outside of a transaction.
+
+.. _mongoid-txn-low-level:
+
+Lower Level API
+---------------
+
+When using the lower level API, you must create a session before
+starting a transaction. You can create a session by calling the
+``with_session()`` method on a model class or an instance of a model.
+
+Then, you can start a transaction by calling the ``start_transaction()``
+method on a session, as shown in the following code:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-lower-lvl-api
+   :end-before: end-lower-lvl-api
+   :language: ruby
+   :dedent:
+
+You can specify a read concern, write concern or read
+preference when starting a transaction by passing options to the
+``start_transaction()`` method:
+
+.. code-block:: ruby
+
+   session.start_transaction(
+       read_concern: {level: :majority},
+       write_concern: {w: 3},
+       read: {mode: :primary}
+   )
+
+When using this API, you must manually commit or end the transaction.
+You can use the ``commit_transaction()`` and ``abort_transaction()``
+methods on the session instance to manage the transaction lifecycle:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-commit-abort
+   :end-before: end-commit-abort
+   :language: ruby
+   :dedent:
+
+.. note::
+
+   If a session ends and includes an open transaction, the transaction is
+   automatically ended.
+
+You can retry the transaction commit if it fails,
+as shown in the following code: 
```

Comment:
S: brief description that the transaction commit is retried on ``UNKNOWN_TRANSACTION_COMMIT_RESULT_LABEL`` exception